snooty/util.py

```diff
@@ -623,3 +632,43 @@ def lines_contain(haystack: Iterable[str], needle: str) -> Iterator[int]:
     of the line where the match succeeded. Repeat for each line."""
     pat = re.compile(rf"^\W*{re.escape(needle)}\W*$")
     yield from (idx for idx, line in enumerate(haystack) if pat.match(line))
+
+
+def structural_hash(obj: object) -> bytes:
+    """Compute a hash of a nested set of dataclasses and primitive types. We form a kind of simple
+    serialization format in the process -- it's just here to prevent ambiguity. Structural hashes
+    are subject to change and are not stable across releases.
+
+    Fields in dataclasses with ma metadata value of "nohash" are skipped."""
+    hasher = hashlib.blake2b()
+    if isinstance(obj, (int, str, float, PurePath)):
+        hasher.update(bytes("P" + str(obj), "utf-8"))
+    elif dataclasses.is_dataclass(obj):
+        fields = dataclasses.fields(obj)
+        hasher.update(bytes(f"O{len(fields)}\x20", "utf-8"))
+        for field in sorted(fields, key=lambda x: x.name):
+            if not field.metadata.get("nohash"):
+                hasher.update(bytes(f"F{len(field.name)}\x20{field.name}", "utf-8"))
+                hasher.update(structural_hash(getattr(obj, field.name)))
+    elif isinstance(obj, (collections.abc.Sequence, collections.abc.Set)):
+        hasher.update(bytes(f"L{len(obj)}\x20", "utf-8"))
```

Comment:
they're type markers to prevent ambiguity in the hash. To be honest ideally we would actually say hash a BSON serialized representation, but that would be more involved. Still, it might be worth doing down the road.